runtime/lua/vim/lsp/callbacks.lua

```diff
@@ -89,6 +92,8 @@ M['textDocument/publishDiagnostics'] = function(_, _, result)
   -- In particular, this stops a ton of spam when first starting a server for current
   -- unloaded buffers.
   if not api.nvim_buf_is_loaded(bufnr) then
+    -- since buffer was loaded by uri_to_bufnr, wipe it out to leave a clean buffer list
+    vim.cmd(string.format('%dbw!', bufnr))
```

Comment:
Is this required w/ the change you made above?

Reply:
It looks unrelated.

Reply:
yep, the first line only prevents buffers with no diagnostics from being loaded. this other line closes the ones that were not loaded by the user (`nvim_buf_is_loaded`). So for example, given the above configuration for Sumneko, all your lua config files and all lua files in `VIMRUNTIME/lua` are processed, many will have no diagnostics and will make the callback return on the new first line. But other will have diagnostics and will be processed, remaining as loaded unlisted buffers. The point is, if the user opens a single lua file, why should it have dozens of unlisted buffers of non-related files just because they have some diagnostics and happened to be processed by sumneko.

 I use a different LSP server with similar behavior. All the ones using LSP workspaces (similar to VisualCode ones), will suffer from this, since the LSP server will process all the files in the workspace and nvim will add a buffer for all of them containing diagnostics.

You can try it on your own. Open a single lua file and then do `ls!`

Reply:
If you wait one or two days, we can use the new API in https://github.com/neovim/neovim/pull/13077 :) 

However, I'm not 100% confident that we actually want to do this... especially given something like this: https://github.com/neovim/neovim/pull/13103 which I think we should merge.

Perhaps a better change would be in vim.uri_to_bufnr, perhaps we can do something else instead of using `vim.fn.bufadd`. Maybe a better option would be adding a new `nvim_add_buf`...? And we could make it so that we don't add that buffer to the buflist when we create it? I'm not 100% sure.